code/modules/power/engines/tesla/tesla_zap.dm

```diff
@@ -0,0 +1,166 @@
+/*
+Alright so, this is a file that has the proc for a proc that works like 90% of all zaps.
+basic_tesla_zap() can be called by anything and does not need any overrides.
+*/
+
+//Zap constants, speeds up targeting
+#define COIL (ROD + 1)
+#define ROD (RIDE + 1)
+#define RIDE (LIVING + 1)
+#define LIVING (MACHINERY + 1)
+#define MACHINERY (BLOB + 1)
+#define BLOB (STRUCTURE + 1)
+#define STRUCTURE (1)
+
+
+// If you want a normal zap, you use the proc below
+
+/proc/basic_tesla_zap(atom/source, zap_range = 3, power, zap_flags = ZAP_DEFAULT_FLAGS, list/shocked_targets = list())
+	if(QDELETED(source))
+		return
+	if(!(zap_flags & ZAP_ALLOW_DUPLICATES))
+		LAZYSET(shocked_targets, source, TRUE) //I don't want no null refs in my list yeah?
+	. = source.dir
+	if(power < 1000)
+		return
+
+	/*
+	THIS IS SO FUCKING UGLY AND I HATE IT, but I can't make it nice without making it slower, check*N rather then n. So we're stuck with it.
+	*/
+	var/atom/closest_atom
+	var/closest_type = 0
+	var/static/things_to_shock = typecacheof(list(/obj/machinery, /mob/living, /obj/structure, /obj/vehicle))
+	var/static/blacklisted_tesla_types = typecacheof(list(/obj/machinery/atmospherics,
+										/obj/machinery/atmospherics/portable,
+										/obj/machinery/power/emitter,
+										/obj/machinery/field/generator,
+										/mob/living/simple_animal/slime,
+										/obj/machinery/particle_accelerator/control_box,
+										/obj/structure/particle_accelerator/fuel_chamber,
+										/obj/structure/particle_accelerator/particle_emitter/center,
+										/obj/structure/particle_accelerator/particle_emitter/left,
+										/obj/structure/particle_accelerator/particle_emitter/right,
+										/obj/structure/particle_accelerator/power_box,
+										/obj/structure/particle_accelerator/end_cap,
+										/obj/machinery/field/containment,
+										/obj/structure/disposalpipe,
+										/obj/structure/disposaloutlet,
+										/obj/machinery/disposal/deliveryChute,
+										/obj/machinery/camera,
+										/obj/structure/sign,
+										/obj/structure/lattice,
+										/obj/structure/grille,
+										/obj/structure/cable,
+										/obj/machinery/the_singularitygen/tesla,
+										/obj/machinery/constructable_frame/machine_frame))
+
+	for(var/a in typecache_filter_multi_list_exclusion(oview(zap_range + 2, source), things_to_shock, blacklisted_tesla_types))
```

Comment:
it looks like the only differences between the new basic_tesla_zap and the tesla_zap method on the energy ball is:
- this one uses oview instead of range
- the energy ball includes APCs as a potential shock target

if you abstracted out this part of the method, this for loop where the shock target is selected, you could avoid having the duplication of most of the tesla_zap proc